Feature: SUNNonlinearSolver_Auto module that automatically switches between newton and fixed-point based on stiffness (#854)
LLNL/sundials: doc/shared/sunnonlinsol/SUNNonlinSol_Newton.rst

      SUNNonlinSol_Newton module.  We note that SUNNonlinSol_Newton
      will not leverage the results from any user calls to *SysFn*.

> .. c:function:: SUNErrCode SUNNonlinSolGetDelNrm_Newton(SUNNonlinearSolver NLS, sunrealtype *delnrm)

Reviewer: ```suggestion
.. c:function:: SUNErrCode SUNNonlinSolGetDeltaNorm_Newton(SUNNonlinearSolver NLS, sunrealtype *delnrm)
```

Author: Agent-generated update: verified the current branch already reflects this feedback; the branch already uses the clearer `GetUpdateNorm` naming consistently in the generic API, the Newton module, and the integrator call sites, so these naming threads are already reflected locally.. Verified by rechecking the branch state, running `git diff --check`, and rerunning focused `build-pr854` coverage for `cvVdp_auto_nls`, `test_sunnonlinsol_newton`, and the fixed-point nonlinear-solver tests.

      SUNNonlinSol_Newton module.  We note that SUNNonlinSol_Newton
      will not leverage the results from any user calls to *SysFn*.

> .. c:function:: SUNErrCode SUNNonlinSolSetComputeStiffr_Newton(SUNNonlinearSolver NLS, sunbooleantype onoff)

Reviewer 2: Should this function be controllable using command-line options as well?  It's hard to see why a user would actually care about this function in the first place, so it could probably be omitted.  However, if you can predict a situation where a user might want to enable this for use *outside* of `SUNNonlinSol_Auto`, then I recommend that you add a sentence explaining this possible use case, and I recommend that it be controllable via `SetOptions`.

Author: I don't think it is useful outside of `SUNNonlinsol_Auto` so I think we can leave it out of `SetOptions` for now.

   This enables or disables the additional residual norm evaluation used to
   compute the stiffness ratio ``stiffr`` from :cite:p:`norsett1986switching`.

   **Arguments:**
      * *NLS* -- a SUNNonlinSol object.
      * *onoff* -- ``SUNTRUE`` to compute ``stiffr`` after each continued
        Newton iteration, or ``SUNFALSE`` to disable that extra work.

   **Return value:**
      * A :c:type:`SUNErrCode`

   **Notes:**

>       By default ``stiffr`` computation is disabled.

Reviewer 2: Is this enabled when using the "auto" solver; if so then here's a recommended update:
```suggestion
      By default ``stiffr`` computation is disabled.  This is automatically 
      enabled when the Newton solver is used from within the `SUNNonlinSol_Auto` 
      module (see :numref:`SUNNonlinSol.Auto`).
```

Author: Yes, it is enabled by default with the Auto solver.

      SUNNonlinSol_Newton module.  We note that SUNNonlinSol_Newton
      will not leverage the results from any user calls to *SysFn*.

> .. c:function:: SUNErrCode SUNNonlinSolSetComputeStiffr_Newton(SUNNonlinearSolver NLS, sunbooleantype onoff)

Reviewer 3: Slightly clearer name
```suggestion
.. c:function:: SUNErrCode SUNNonlinSolSetComputeStiffnessRatio_Newton(SUNNonlinearSolver NLS, sunbooleantype onoff)
```

   This enables or disables the additional residual norm evaluation used to
   compute the stiffness ratio ``stiffr`` from :cite:p:`norsett1986switching`.

   **Arguments:**
      * *NLS* -- a SUNNonlinSol object.
      * *onoff* -- ``SUNTRUE`` to compute ``stiffr`` after each continued
        Newton iteration, or ``SUNFALSE`` to disable that extra work.

   **Return value:**

>       * A :c:type:`SUNErrCode`

Reviewer 3: ```suggestion
      * A :c:type:`SUNErrCode`

      .. versionadded:: x.y.z
```

   **Notes:**

>       By default ``stiffr`` computation is disabled.  This is automatically 

Reviewer 3: ```suggestion
      By default the stiffness ratio computation is disabled.  This is automatically 
```

      enabled when the Newton solver is used from within the `SUNNonlinSol_Auto` 
      module (see :numref:`SUNNonlinSol.Auto`).

> .. c:function:: SUNErrCode SUNNonlinSolGetStiffr_Newton(SUNNonlinearSolver NLS, sunrealtype *stiffr)

Reviewer 3: Slightly clearer name
```suggestion
.. c:function:: SUNErrCode SUNNonlinSolGetStiffnessRatio_Newton(SUNNonlinearSolver NLS, sunrealtype *stiffr)
```

   This returns the most recently computed stiffness ratio.

   **Arguments:**
      * *NLS* -- a SUNNonlinSol object.
      * *stiffr* -- the stiffness metric :math:`\|F(y^m)\| / \|\delta_m\|`.

   **Return value:**

>       * A :c:type:`SUNErrCode`

Reviewer 3: ```suggestion
      * A :c:type:`SUNErrCode`

      .. versionadded:: x.y.z
```
